source/robot_model/src/Model.cpp

```diff
@@ -260,9 +263,114 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
+      if (this->robot_model_.upperPositionLimit[n] < joint_positions.get_positions()[n]) {
+        W_b(n, n) = 0;
+      } else {
+        d = (joint_positions.get_positions()[n] - (this->robot_model_.upperPositionLimit[n] - margin)) / margin;
+        W_b(n, n) = -2 * d * d * d + 3 * d * d;
+      }
+    }
+  }
+  return W_b;
+}
+
+Eigen::VectorXd Model::cwln_repulsive_potential_field(const state_representation::JointPositions& joint_positions, double margin) {
+  Eigen::VectorXd Psi(this->robot_model_.nq);
+  Eigen::VectorXd q = joint_positions.get_positions();
+  for (int i = 0; i < this->robot_model_.nq; i++) {
+    Psi[i] = 0;
+    if(q[i] < this->robot_model_.lowerPositionLimit[i] + margin) {
+      Psi[i] = this->robot_model_.upperPositionLimit[i] - margin - std::max(q[i], this->robot_model_.lowerPositionLimit[i]);
+    } else if(this->robot_model_.upperPositionLimit[i] - margin < q[i]) {
+      Psi[i] = this->robot_model_.lowerPositionLimit[i] + margin - std::min(q[i], this->robot_model_.upperPositionLimit[i]);
+    }     
+  }
+  
+  return Psi;
+}
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                              const state_representation::JointState& current_joint_state,
+                                                              std::string frame_name,
+                                                              const Model::InverseGeometryParameters& params) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  } else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) {
+      throw (exceptions::FrameNotFoundException(frame_name));
+    }
+  }
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  // 1 second for the Newton-Raphson method
+  const std::chrono::nanoseconds dt((int)(1e9));
+
+  state_representation::JointPositions q(current_joint_state);
+  state_representation::JointVelocities dq(this->get_robot_name(), this->get_number_of_joints());
+  state_representation::Jacobian J(this->get_robot_name(), this->robot_model_.nq, frame_name);
+  Eigen::MatrixXd J_b = J.data();
+  Eigen::MatrixXd JJt(6,6);
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::MatrixXd W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::VectorXd psi(this->robot_model_.nq);
+  Eigen::VectorXd err(6);
+  state_representation::CartesianPose X(this->get_robot_name());
+  const state_representation::CartesianPose X_d(desired_cartesian_state);
+  
+  int i = 0;
+  while (i < params.max_number_of_iterations) {
+    X = this->forward_geometry(q, frame_id);
+    
+    err = ((X_d - X)/dt).data();
```

Comment:
Couldn't you just do `(X_d - this->forward_geometry(q, frame_id))` to get rid of `X` ?